### ISSUE TITLE
Race condition between drain close and integration server shutdown

### ISSUE THREAD

**Issue by auni53** [envoyproxy/envoy#11240]
I was running into weird time behaviour as I was trying to add more integration test coverage for drain closing behaviour. I took out the simulated time, took out the WIP changes in review, and isolated it to just the drain timeout plumbing on the original protocol integration test.

https://github.com/auni53/envoy/pull/3

The first comment is a "good run", `drain_time_ = std::chrono::seconds(1);`, with trace comments enabled. the second comment has `drain_time_ = std::chrono::seconds(10)` and hangs infinitely on the stat flushing loop. I'm not really familiar enough with the shutdown sequence to understand what's not being triggered. I got tsan to find some errors yesterday but can't seem to be able to re-trigger that.

**Comment by auni53** [comment]
@mattklein123 if you could take a look that would be appreciated!

**Comment by auni53** [comment]
Got tsan to pick up again! Posted a third comment on that draft PR, I'm a little lost on interpreting what tsan is saying there.

**Comment by mattklein123** [comment]
The issuing I'm assuming is that when you increase the drain time it's not deterministic as to whether the request gets drained closed and that's why it hangs? The test probably needs to happen in a more deterministic way.

**Comment by auni53** [comment]
OK, many many painful iterations, I think I got to the bottom of this. I stripped out the graceful admin work, the simulated time framework, reset to master, and finally was able to reproduce are as follows:
1. Allow drain time to be configured in the integration tests
2. By setting `drain_time_ = x` in [ProtocolIntegrationTest.DrainClose](https://github.com/envoyproxy/envoy/blob/master/test/integration/protocol_integration_test.cc#L278), the probability of test failure increases with x.

The question remained, why? I was kind of stumped for why the server was hanging, and tsan wasn't very intelligible. But I tried reverting #11176, where I removed the TestDrainManager, and the tests started passing. So I started dumping logs and trying to compare the pass and fail cases, and incrementally mutating TestDrainManager to be more like DrainManagerImpl, and mutating DrainManagerImpl to be like TestDrainManager to find the root cause. Finally found that by sticking a `return true;` before the DrainManagerImpl [probabilistic drain close logic](https://github.com/envoyproxy/envoy/blob/master/source/server/drain_manager_impl.cc#L39), the problem goes away.

So basically, on stream end, the conn manager asks its DrainDecision whether we're in the drain sequence. Due to the probabilistic drain close, this was usually returning false, resulting in no HTTP2 GOAWAY being sent. Therefore, the stream would complete, the connection would stay open, and the test hangs on the [codec_client_->WaitForDisconnect()](https://github.com/envoyproxy/envoy/blob/master/test/integration/protocol_integration_test.cc#L292). I'm speaking to Matt in a bit to figure out next steps.

**Comment by mattklein123** [comment]
cc @alyssawilk also who I think has been also dealing with a similar issue trying to speed up some of the integration tests.

**Comment by auni53** [comment]
Chatted with Alyssa, her work in #11260 seems to be unrelated.

Action items to resolve this:

1. Try to deflake the test by using retries.
2. Add a config knob to drain manager to disable probabilistic drain close, which will disable the race.
3. Move the test over to simulated time, and just jump forward past the probabilistic drain period.
